### PR TITLE
compressor/lz4: work around bug in liblz4 versions <1.8.2

### PR DESCRIPTION
**Pull request by liewegas**
In older versions of lz4 (anything older than 1.8.2) bit errors can be introduced when compressing from fragmented source memory (i.e., a sequence of non-continguous memory extents).

This is a workaround to try to avoid situations that would trigger the bug.  liblz4 doesn't have a runtime version check, so we have to do something for all versions (even new versions) and hopefully avoid impacting performance.

Fixes: https://tracker.ceph.com/issues/39525 

Also includes a unit test that reproduces the original issue, and expanded coverage of compressor algorithms for the qa suite.

### CODE REVIEW

**Comment by liewegas** [comment]
@dvanders

File: src/test/compressor/test_compression.cc [hunk 1]
```diff
@@ -330,9 +369,9 @@ INSTANTIATE_TEST_SUITE_P(
     "lz4",
 #endif
 #ifdef __x86_64__
-    "zlib/isal",
+//    "zlib/isal",
```
Comment (dvanders) [inline src/test/compressor/test_compression.cc]: Oops -- don't need to disable these.

File: src/test/compressor/test_compression.cc [hunk 2]
```diff
@@ -330,9 +369,9 @@ INSTANTIATE_TEST_SUITE_P(
...
 #endif
-    "zlib/noisal",
+//    "zlib/noisal",
```
Comment (dvanders) [inline src/test/compressor/test_compression.cc]: Same as above.

**Comment by liewegas** [comment]
retest this please

**Comment by dvanders** [comment]
@cyan4973 hope you don't mind I ping you like this.
We found this lz4 bit flip issue in Ceph with a particular bufferlist which was not word aligned. Does it ring a bell?

**Comment by Cyan4973** [comment]
Wow, nope, that's new to me.

That's strange. Any issue with regards to unaligned memory access is presumed already fixed before `v1.7.5` .

Things that could be attempted : 
- test different values of `LZ4_FORCE_MEMORY_ACCESS`, from `0` (most portable) to `2` (most aggressive)
- check `-O2`, see if it differs from `-O3`
- for `gcc` : add flag `-fno-strict-aliasing`

As mentioned, these potential issues, related to aggressive compiler optimizations, were supposed to be fixed before `v1.7.5`, but well, maybe one was still lurking in a corner ....

File: src/test/compressor/test_compression.cc [hunk 3]
```diff
@@ -157,6 +158,44 @@ TEST_P(CompressorTest, big_round_trip_file)
...
+    //fbl.rebuild();
+    bufferlist compressed;
+    int r = compressor->compress(chunk, compressed);
+    ASSERT_EQ(0, r);
+    bufferlist decompressed;
+    r = compressor->decompress(compressed, decompressed);
+    ASSERT_EQ(0, r);
+    ASSERT_EQ(decompressed.length(), chunk.length());
+    if (!decompressed.contents_equal(chunk)) {
+      cout << "FAILED, orig bl was\n" << fbl << std::endl;
```
Comment (dvanders) [inline src/test/compressor/test_compression.cc]: nit: "orig bl was" -> "fbl was"

**Comment by dvanders** [comment]
Thanks @Cyan4973 for having a look.
On my machine (lz4-1.7.5-3.el7.x86_64) I get the corruption for any `LZ4_FORCE_MEMORY_ACCESS` from `0` to `2`. This log shows the bufferlist we are compressing: https://termbin.com/qpcl

AFAICT we already use `-fno-strict-aliasing`: https://github.com/ceph/ceph/blob/mimic/src/CMakeLists.txt#L32

For `-O2|3` I presume you mean recompile lz4 ?

**Comment by liewegas** [comment]
I think it would be illuminating to build lz4 from source and see (1) if the build options affect behavior, and (2) in which version the problem is fixed...

**Comment by liewegas** [comment]
Using default build options, this appears to be fixed in v1.8.2.  v1.8.1.2 is still broken.

**Comment by liewegas** [comment]
-O2 and -O1 have no effect.

**Comment by liewegas** [comment]
LZ4_FORCE_MEMORY_ACCESS 0 to 2 also has no effect (on 1.7.5).

**Comment by liewegas** [comment]
I bisected it and this is the commit that fixes it:
```
commit af127334670a5e7b710bbd6adb71aa7c3ef0cd72
Author: Yann Collet <cyan@fb.com>
Date:   Sat May 5 18:24:11 2018 -0700

    fixed frametest error
    
    The error can be reproduced using following command :
    ./frametest -v -i100000000 -s1659 -t31096808
    
    It's actually a bug in the stream LZ4 API,
    when starting a new stream
    and providing a first chunk to complete with size < MINMATCH.
    In which case, the chunk becomes a dictionary.
    No hash was generated and stored,
    but the chunk is accessible as default position 0 points to dictStart,
    and position 0 is still within MAX_DISTANCE.
    Then, next attempt to read 32-bits from position 0 fails.
    
    The issue would have been mitigated by starting from index 64 KB,
    effectively eliminating position 0 as too far away.
    
    The proper fix is to eliminate such "dictionary" as too small.
    Which is what this patch does.
```

**Comment by liewegas** [comment]
@Cyan4973 Can you give us some guidance as to whether the workaround (consolidating to a single, contiguous buffer if any parts of the stream are not word-aligned) is an appropriate workaround for this bug?  Thanks!

**Comment by Cyan4973** [comment]
Thanks for bisection @liewegas !

This fix is applicable to scenarios requesting LZ4 to build several separated blocks, from data scattered in memory, while still maintaining history between them. This is a pretty advanced scenario.
It also happens when doing some dictionary compression.

Strangely enough, it does not seem to be related to unaligned accesses.

Consolidating input as a single contiguous buffer is bound to improve the situation. If my understanding is correct, it means providing the whole input as a single block to compress ? Indeed, this is much easier for the compressor, and is very well tested. Performance (both speed and compression ratio) should improve too.

**Comment by liewegas** [comment]
@Cyan4973 I've updated 3fce46a891dcb275d2bafb24ba02b64b34d4117f to rebuild into a contiguous buffer if the source is not already composed of page-aligned and page-sized fragments.  Is this sufficient to work around this bug?  I.e., could the problem still trigger with a series for 4096-byte pieces?

We could unconditionally rebuild the buffer into contiguous memory, but I'm a bit worried about the performance impact.  (Although, tbh, I think that in most cases the source material will already be a single buffer.)

**Comment by liewegas** [comment]
@Cyan4973 I've just updated to always compress a contiguous buffer.  Should be simpler, safer, and faster!

**Comment by Cyan4973** [comment]
Yes, this is totally helpful

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2020-03-03_13:02:55-rados-wip-kefu-testing-2020-03-03-1622-distro-basic-smithi/

the failure were either 

- caused by cephadm 
- fixed by #33701

**Pull request merged**